### PR TITLE
Add logger flag to save_hyperparameters

### LINKED ISSUE #6338: Support save_hyperparameters() to checkpoints without sending to logger
## 🚀 Feature
`self.save_hyperparameters()` is an awesome simple way to save input arguments with checkpoints such that when loading a checkpoint the module will be constructed with the same parameters. The method also seems to log the hyperparameters to the experiment logger, which is not always desired. It would be great to add a flag to turn this on or off.

### Pitch

Add a `log=True` flag, which is true by default, to `save_hyperparameters()`. This way hyperparameters could be saved but not logged by calling `self.save_hyperparameters(log=False)`.

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #6338 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -214,6 +215,14 @@ def logger(self):
         """ Reference to the logger object in the Trainer. """
         return self.trainer.logger if self.trainer else None
 
+    @property
+    def log_hyperparams(self) -> bool:
+        return self._log_hyperparams
+
+    @log_hyperparams.setter
+    def log_hyperparams(self, log: bool) -> None:
+        self._log_hyperparams = log
```
Comment (justusschock) [inline pytorch_lightning/core/lightning.py:224]: why do we need a property for that if it only forwards it to a protected attribute?

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1738,7 +1747,8 @@ def save_hyperparameters(
         self,
         *args,
         ignore: Optional[Union[Sequence[str], str]] = None,
-        frame: Optional[types.FrameType] = None
+        frame: Optional[types.FrameType] = None,
+        log: bool = True
```
Comment (justusschock) [inline pytorch_lightning/core/lightning.py:1751]: docstrings are missing

**Review by justusschock — COMMENTED** [review]
Did you try this with tensorboard? I think we disabled this on purpose, since actually we had issues that if we log them in the beginning, TB (the most common logger and default) does not properly update the metrics.

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:1751]: Let's be consistent and use `logger: bool = True`

Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py]: Done!

Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py]: Done!

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py]: Updated!

File: CHANGELOG.md [hunk 3]
```diff
@@ -102,7 +102,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added support `LightningModule.save_hyperparameters` when `LightningModule` is a dataclass ([#7992](https://github.com/PyTorchLightning/pytorch-lightning/pull/7992))
 
 
-- Add support for overriding `optimizer_zero_grad` and `optimizer_step` when using accumulate_grad_batches ([#7980](https://github.com/PyTorchLightning/pytorch-lightning/pull/7980))
+- Added support for overriding `optimizer_zero_grad` and `optimizer_step` when using accumulate_grad_batches ([#7980](https://github.com/PyTorchLightning/pytorch-lightning/pull/7980))
+
+
+- Added `log` flag to `save_hyperparameters` ([#7960](https://github.com/PyTorchLightning/pytorch-lightning/pull/7960))
```
Comment (awaelchli) [inline CHANGELOG.md:108]:
```suggestion
- Added `logger` boolean flag to `save_hyperparameters` ([#7960](https://github.com/PyTorchLightning/pytorch-lightning/pull/7960))
```
mention boolean just to make sure it is not misunderstood as logger instance

File: tests/loggers/test_base.py [hunk 4]
```diff
@@ -290,3 +291,37 @@ def log_hyperparams(self, params):
...
+def test_log_hyperparams_being_called(tmpdir, logger):
+
+    class TestLogger(DummyLogger):
+
+        def __init__(self):
+            super().__init__()
+            self.log_hyperparams_called = False
+
+        def log_hyperparams(self, *args, **kwargs):
+            self.log_hyperparams_called = True
```
Comment (awaelchli) [inline tests/loggers/test_base.py:306]: can you patch the method in the logger instead?

With the decorator
```python
@patch("pytorch_lightning.loggers.tensorboard.TensorBoard.log_hyperparams")
```
or passing a mock as the logger object

```python
logger_mock = Mock()
trainer = Trainer(logger=logger_mock, ...)
trainer.fit(...)

mock.log_hyperparams.assert_not_called()
```

File: pytorch_lightning/utilities/hparams_mixin.py [hunk 5]
```diff
@@ -30,7 +30,8 @@ def save_hyperparameters(
         self,
         *args,
         ignore: Optional[Union[Sequence[str], str]] = None,
-        frame: Optional[types.FrameType] = None
+        frame: Optional[types.FrameType] = None,
+        logger: bool = True
```
Comment (ananthsub) [inline pytorch_lightning/utilities/hparams_mixin.py:34]: similarly why does the logger need to be added here?

File: pytorch_lightning/core/datamodule.py [hunk 6]
```diff
@@ -97,6 +97,8 @@ def __init__(
         self._has_teardown_test = False
         self._has_teardown_predict = False
 
+        self._log_hyperparams = True
```
Comment (ananthsub) [inline pytorch_lightning/core/datamodule.py:100]: - should this be part of the hyperparams mixin? 
- why is it protected? are users expected to toggle this on or off?

File: tests/loggers/test_base.py (hunk 4 above)
Comment (kaushikb11) [reply tests/loggers/test_base.py]: @awaelchli Updated!

File: pytorch_lightning/utilities/hparams_mixin.py (hunk 5 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/hparams_mixin.py:34]: >similarly why does the logger need to be added here? should we add a separate property like should_log_hyperparams to decouple this?

Tbh, I like the interface where Users could define in `save_hyperparameters` itself if they wish to log the hyperparameters or not.

File: pytorch_lightning/core/datamodule.py (hunk 6 above)
Comment (kaushikb11) [reply pytorch_lightning/core/datamodule.py:100]: > should this be part of the hyperparams mixin?

Could make it work. Would need an additional check for `_log_hyperparams` in Trainer.
```python
if not hasattr(self, "_log_hyperparams"):
```

> why is it protected? are users expected to toggle this on or off?

Nope, should be expected to toggle with the `save_hyperparameters` api.

Comment (awaelchli) [reply pytorch_lightning/core/datamodule.py:100]: yes this should go to the mixin class, otherwise the LightningModule has a dependency on the Mixin interface, but this should not be, or we can't call it a mixin.  

we can simply add the initialization of this attribute to the mixin, right?

Comment (kaushikb11) [reply pytorch_lightning/core/datamodule.py]: Updated!

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/datamodule.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/core/datamodule.py]: eggsellent!

**Review by SeanNaren — APPROVED** [review]
nice feature!

File: pytorch_lightning/utilities/hparams_mixin.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/utilities/hparams_mixin.py:38]: Yes, IMO I think this is quite a neat API.

**Review by tchaton — APPROVED** [review]
Love it !
